msprime/intervals.py

```diff
@@ -33,83 +33,232 @@ class RateMap:
+        chromosome (often the telomeric region). It primarily affects the
+        :attr:`mean_rate` returned for this rate map.
+        """
+        return self._end_position
+
+    @end_position.setter
+    def end_position(self, val):
+        raise NotImplementedError(
+            "Cannot change a RateMap end_position: use ratemap.slice(... trim=False)"
+        )
+
     @staticmethod
     def uniform(sequence_length, rate):
+        """
+        Create a uniform rate map
+        """
         return RateMap([0, sequence_length], [rate])
 
     def asdict(self):
         return {"position": self.position, "rate": self.rate}
 
     @property
     def sequence_length(self):
+        """
+        The sequence length covered by this map
+        """
         return self.position[-1]
 
     @property
     def total_mass(self):
-        return self.cumulative[-1]
+        """
+        The cumulative total over the entire map
+        """
+        # Since we insist that the rate up to start_position and past end_position is 0
+        # we can just return the cumulative total.
+        return self.cumulative_mass[-1]
 
     @property
-    def size(self):
+    def size(self):  # TODO - should we remove this, for the same reason as for __len__
         return self.rate.shape[0]
 
     @property
     def mean_rate(self):
         """
-        Return the weighted mean across all windows of the entire map.
+        The mean rate over this map, from :attr:`start_position` to
+        :attr:`end_position`, weighed by the span covered by each rate.
         """
-        window_sizes = self.position[1:] - self.position[:-1]
-        weights = window_sizes / self.sequence_length
-        if self.map_start != 0:
-            weights[0] = 0
-        return np.average(self.rate, weights=weights)
+        span = self.end_position - self.start_position
+        return self.total_mass / span
 
-    def slice(self, start=None, end=None, trim=False):  # noqa: A003
+    def get_rate(self, x):
         """
-        Returns a subset of this rate map between the specified end
-        points. If start is None, it defaults to 0. If end is None, it defaults
-        to the end of the map. If trim is True, remove the flanking
-        zero rate regions such that the sequence length of the
-        new rate map is end - start.
+        Return the rate at a list of user-specified positions along the map, i.e.
+        interpolated between the fixed :attr:`position` values. Any positions
+        beyond the end  of the map will return the final rate value.
+
+        :param numpy.ndarray x: The positions for which to return values.
+
+        :return: An array of rates, the same length as ``x``.
+        :rtype: numpy.ndarray(dtype=np.float64)
+        """
+        loc = np.searchsorted(self.position, x, side="right") - 1
+        if np.any(loc < 0):
+            raise ValueError("No rate exists for positions below the minimum map pos")
```

Comment:
Took me a minute to figure out what this was doing, this would be more obvious:
```python
loc = np.minimum(loc, len(self.rate) - 1)
return self.rate[loc]
```